File: bench/loomyard-eval/ladder/scripts/summarize.py
"""
Summarises the quarry-mcp capability ladder's completed matrix into the
tracked `summary.json`: per-config medians and full ranges over every
metric `extract_usage.py` and `score_run.py` write, and the disjoint-range
separation rule applied independently to the three comparison types the
discussion defines -- rung-vs-control, rung-vs-rung, and warm-vs-cold.

Completeness is never re-derived here: `load_runs` calls `gates.is_complete`
so there is exactly one definition of a complete run across the suite.

Usage:
    from summarize import build_summary, write_summary
    write_summary(ladder, results_root)
"""
import argparse
import json
import sys
from dataclasses import asdict, dataclass
from pathlib import Path

from gates import is_complete, run_dir
from ladder_config import config_by_id, control_for, load_ladder, warm_counterpart_for

# Ordered metric names summarised into a median/min/max/n record. Drawn from
# every run's usage.json and score.json, except decoy_admitted and
# summary_matches, which are qualitative or boolean and get their own
# per-cell treatment in build_summary instead of a meaningless median.
METRICS = (
    "duration_ms",
    "wall_clock_ms",
    "num_turns",
    "tool_uses",
    "quarry_tool_uses",
    "input_tokens",
    "output_tokens",
    "cache_read_input_tokens",
    "cache_creation_input_tokens",
    "cost_usd",
    "bash_grep_count",
    "grep_tool_count",
    "grep_fallback_total",
    "denied_tool_attempts",
    "recall",
    "precision",
    "lookalikes_matched",
)

# The subset of METRICS excluded from a rung-vs-control comparison: the
# control's preamble differs in steering as well as tools, so a grep-usage
# gap between it and a rung cannot be attributed to the tool exposure alone.
GREP_METRICS = ("bash_grep_count", "grep_tool_count", "grep_fallback_total")

# run.json observations load_runs lifts onto each per-run mapping, recorded
# by the harness as non-fatal gate findings rather than summarised metrics.
_RUN_JSON_OBSERVATIONS = ("worktree_dirtied", "target_origin_quarry_mention", "cold_no_daemon_backed_call")


class SummarizeError(Exception):
    """Raised when a comparison is requested across configs that cannot be
    compared -- e.g. compare_rungs given two configs from different
    ladders."""


""" PER-CONFIG MEDIANS, RANGES, AND COMPLETENESS """


@dataclass(frozen=True)
class Cell:
    """One config's summarised results.

    Instance variables:
        config_id: the LadderConfig id this cell belongs to.
        runs: the list of flattened per-run mappings load_runs read.
        complete: True only when len(runs) == the config's reps.
        stats: mapping from METRICS name to a record of median/min/max/n,
            present only for metrics carried by at least one run.
    """

    config_id: str
    runs: list
    complete: bool
    stats: dict


def load_runs(results_root, config_id, reps):
    """
    Reads every complete run of config_id under
    <results_root>/raw/<config_id>/<n>/ for n in 1..reps.

    A run directory is skipped when gates.is_complete returns False for it
    -- this is the suite's single definition of completeness, never
    re-derived here. For each complete run, usage.json, score.json, and
    run.json are merged into one flat mapping: usage.json's tokens.<class>
    fields are lifted to top-level <class> keys, every other usage.json and
    score.json field is copied unchanged, and run.json's worktree_dirtied,
    target_origin_quarry_mention, and cold_no_daemon_backed_call
    observations are carried across when present.

    Returns:
        A list of flattened per-run mappings, in repetition order, one per
        complete run found.
    """
    runs = []
    for n in range(1, reps + 1):
        a_run_dir = run_dir(results_root, config_id, n)
        if not is_complete(a_run_dir):
            continue

        with open(a_run_dir / "usage.json") as f:
            usage = json.load(f)
        with open(a_run_dir / "score.json") as f:
            score = json.load(f)
        with open(a_run_dir / "run.json") as f:
            run_record = json.load(f)

        merged = {}
        for key, value in usage.items():
            if key == "tokens":
                merged.update(value)
            else:
                merged[key] = value
        merged.update(score)
        for observation in _RUN_JSON_OBSERVATIONS:
            if observation in run_record:
                merged[observation] = run_record[observation]

        runs.append(merged)
    return runs


def _median(values):
    """
    The true median of values: the middle value at odd length, the mean of
    the two middle values at even length.
    """
    ordered = sorted(values)
    count = len(ordered)
    midpoint = count // 2
    if count % 2 == 1:
        return ordered[midpoint]
    return (ordered[midpoint - 1] + ordered[midpoint]) / 2


def summarise_cell(config_id, runs, reps):
    """
    Builds the Cell for config_id from its already-loaded runs.

    A metric absent from a run -- cost_usd when the envelope omitted it, the
    impact-only fields on a Ladder-A run -- is skipped for that run, so its
    stats record's n reflects only the runs that carried it rather than the
    full run count.
    """
    stats = {}
    for metric in METRICS:
        values = [run[metric] for run in runs if metric in run and run[metric] is not None]
        if not values:
            continue
        stats[metric] = {
            "median": _median(values),
            "min": min(values),
            "max": max(values),
            "n": len(values),
        }
    return Cell(config_id=config_id, runs=runs, complete=len(runs) == reps, stats=stats)


""" THE THREE DISJOINT-RANGE COMPARISON TYPES """


@dataclass(frozen=True)
class Comparison:
    """
    One structured, mechanically checkable comparison between two cells over
    one metric.

    Instance variables:
        kind: "rung-vs-control", "rung-vs-rung", or "warm-vs-cold".
        left: the left-hand config id.
        right: the right-hand config id.
        metric: the METRICS name compared.
        left_median, right_median: each side's median.
        left_range, right_range: each side's (min, max) tuple.
        separated: True only when left_range and right_range do not overlap
            at all.
    """

    kind: str
    left: str
    right: str
    metric: str
    left_median: float
    right_median: float
    left_range: tuple
    right_range: tuple
    separated: bool


def ranges_disjoint(a, b):
    """
    True only when the two closed ranges (min, max) do not overlap at all.
    Ranges that touch at a single value -- one's max equals the other's min
    -- are not disjoint.
    """
    a_min, a_max = a
    b_min, b_max = b
    return a_max < b_min or b_max < a_min


def _build_comparison(kind, left_cell, right_cell, metric):
    """Returns one Comparison for metric between left_cell and right_cell,
    or None when either side carries no stats for that metric."""
    left_stats = left_cell.stats.get(metric)
    right_stats = right_cell.stats.get(metric)
    if left_stats is None or right_stats is None:
        return None

    left_range = (left_stats["min"], left_stats["max"])
    right_range = (right_stats["min"], right_stats["max"])
    return Comparison(
        kind=kind,
        left=left_cell.config_id,
        right=right_cell.config_id,
        metric=metric,
        left_median=left_stats["median"],
        right_median=right_stats["median"],
        left_range=left_range,
        right_range=right_range,
        separated=ranges_disjoint(left_range, right_range),
    )


def compare_rung_to_control(ladder, cells, config):
    """
    Compares config against its own ladder's control, resolved through
    control_for rather than by parsing config.id. Emits a comparison for
    every METRICS entry except GREP_METRICS, since the control's preamble
    differs in steering as well as tools. Emits nothing when either cell is
    incomplete.
    """
    control = control_for(ladder, config)
    rung_cell = cells[config.id]
    control_cell = cells[control.id]
    if not rung_cell.complete or not control_cell.complete:
        return []

    comparisons = []
    for metric in METRICS:
        if metric in GREP_METRICS:
            continue
        comparison = _build_comparison("rung-vs-control", rung_cell, control_cell, metric)
        if comparison is not None:
            comparisons.append(comparison)
    return comparisons


def compare_rungs(ladder, cells, left_id, right_id):
    """
    Compares two configs in the same ladder, both with a non-empty allowed
    set expected. All METRICS are eligible, including the grep metrics,
    since these preambles are identical except for the tool list.

    Raises SummarizeError when left_id and right_id are not in the same
    ladder. Emits nothing when either cell is incomplete.
    """
    left_config = config_by_id(ladder, left_id)
    right_config = config_by_id(ladder, right_id)
    if left_config.ladder != right_config.ladder:
        raise SummarizeError(f"compare_rungs: {left_id!r} and {right_id!r} are not in the same ladder")

    left_cell = cells[left_id]
    right_cell = cells[right_id]
    if not left_cell.complete or not right_cell.complete:
        return []

    comparisons = []
    for metric in METRICS:
        comparison = _build_comparison("rung-vs-rung", left_cell, right_cell, metric)
        if comparison is not None:
            comparisons.append(comparison)
    return comparisons


def compare_warm_cold(ladder, cells, cold_disposition):
    """
    Compares the ladder's cold cell against the warm cell named by its own
    warm_counterpart field, resolved through warm_counterpart_for.

    Emits nothing when cold_disposition is "not-run" or "partial" -- a
    disjoint-range claim at n = reps cannot be made from fewer than reps
    cold runs -- when either cell is incomplete, or when every cold run
    recorded cold_no_daemon_backed_call: none of them started a daemon, so
    there is no warmth contrast to draw.
    """
    if cold_disposition in ("not-run", "partial"):
        return []

    cold_config = next(config for config in ladder.configs if config.cold)
    warm_config = warm_counterpart_for(ladder, cold_config)
    cold_cell = cells[cold_config.id]
    warm_cell = cells[warm_config.id]
    if not cold_cell.complete or not warm_cell.complete:
        return []
    if cold_cell.runs and all(run.get("cold_no_daemon_backed_call") for run in cold_cell.runs):
        return []

    comparisons = []
    for metric in METRICS:
        comparison = _build_comparison("warm-vs-cold", cold_cell, warm_cell, metric)
        if comparison is not None:
            comparisons.append(comparison)
    return comparisons


""" THE TRACKED summary.json """


def _read_json_or_default(path, default):
    """Returns the parsed JSON at path, or default when path does not
    exist. Neither missing file this module reads (cold_cell.json,
    probe.json) raises -- summarising a partial matrix is a legitimate
    intermediate state."""
    path = Path(path)
    if not path.exists():
        return default
    with open(path) as f:
        return json.load(f)


def build_summary(ladder, results_root):
    """
    Builds the full mapping written to summary.json: _meta, cells,
    comparisons, and incomplete.

    _meta carries the pinned run model, the pinned scorer mapping, reps,
    the results-root date segment, the number of configs, the cold cell's
    disposition (from cold_cell.json, or "unknown" when absent), and a
    denied_tool_attempts_reported flag (from probe.json's
    denied_tools_advertised key, or None when probe.json is absent).

    cells maps every config id to its Cell stats plus complete,
    decoy_admitted_count, the verbatim summary_matches list,
    worktree_dirtied_count, target_origin_quarry_mention_count, and
    daemon_backed_runs, aggregated from load_runs's per-run observations.

    comparisons is every Comparison the three builders produce, as plain
    dicts. rung-vs-control comparisons are built for every config with a
    non-empty allowed set that is not the cold config; rung-vs-rung
    comparisons are built for every pair of such configs within the same
    ladder; warm-vs-cold is built once, for the ladder's single cold cell.

    incomplete lists the ids of every cell that is not complete, except a
    cold cell whose cold_cell.json records "not-run" or "partial" -- both
    are legitimate terminal states of the cold-cell driver, not interrupted
    runs.
    """
    results_root = Path(results_root)

    cells = {}
    for config in ladder.configs:
        runs = load_runs(results_root, config.id, ladder.reps)
        cells[config.id] = summarise_cell(config.id, runs, ladder.reps)

    cold_cell_record = _read_json_or_default(results_root / "cold_cell.json", {})
    cold_disposition = cold_cell_record.get("disposition", "unknown")
    cold_confirmed_cold_reps = cold_cell_record.get("confirmed_cold_reps")

    probe_record = _read_json_or_default(results_root / "probe.json", None)
    denied_tool_attempts_reported = None if probe_record is None else probe_record.get("denied_tools_advertised")

    meta = {
        "run_model": ladder.run_model,
        "scorer": {"model": ladder.scorer.model, "effort": ladder.scorer.effort},
        "reps": ladder.reps,
        "results_root_date": results_root.name,
        "num_configs": len(ladder.configs),
        "cold_disposition": cold_disposition,
        "cold_confirmed_cold_reps": cold_confirmed_cold_reps,
        "denied_tool_attempts_reported": denied_tool_attempts_reported,
    }

    cell_records = {}
    for config in ladder.configs:
        cell = cells[config.id]
        runs = cell.runs
        cell_records[config.id] = {
            "stats": cell.stats,
            "complete": cell.complete,
            "decoy_admitted_count": sum(1 for run in runs if run.get("decoy_admitted") is True),
            "summary_matches": [run["summary_matches"] for run in runs if "summary_matches" in run],
            "worktree_dirtied_count": sum(1 for run in runs if run.get("worktree_dirtied") is True),
            "target_origin_quarry_mention_count": sum(
                1 for run in runs if run.get("target_origin_quarry_mention") is True
            ),
            "daemon_backed_runs": sum(1 for run in runs if not run.get("cold_no_daemon_backed_call", False)),
        }

    non_control_configs = [config for config in ladder.configs if config.allowed and not config.cold]

    comparisons = []
    for config in non_control_configs:
        comparisons.extend(compare_rung_to_control(ladder, cells, config))
    for ladder_name in ("a", "b"):
        rung_ids = [config.id for config in non_control_configs if config.ladder == ladder_name]
        for i, left_id in enumerate(rung_ids):
            for right_id in rung_ids[i + 1 :]:
                comparisons.extend(compare_rungs(ladder, cells, left_id, right_id))
    comparisons.extend(compare_warm_cold(ladder, cells, cold_disposition))

    incomplete = []
    for config in ladder.configs:
        cell = cells[config.id]
        if cell.complete:
            continue
        if config.cold and cold_disposition in ("not-run", "partial"):
            continue
        incomplete.append(config.id)

    return {
        "_meta": meta,
        "cells": cell_records,
        "comparisons": [asdict(comparison) for comparison in comparisons],
        "incomplete": incomplete,
    }


def write_summary(ladder, results_root):
    """
    Serialises build_summary(ladder, results_root) as summary.json into
    results_root, with sorted keys and a trailing newline.

    Returns the summary mapping, so callers (including the CLI below) do
    not have to re-read the file they just wrote.
    """
    results_root = Path(results_root)
    summary = build_summary(ladder, results_root)
    with open(results_root / "summary.json", "w") as f:
        json.dump(summary, f, indent=2, sort_keys=True)
        f.write("\n")
    return summary


def _exit_code_for_summary(summary):
    """1 when summary's incomplete list is non-empty, 0 otherwise -- a
    summary of a partial matrix is written but must not be mistaken for a
    finished one."""
    return 1 if summary["incomplete"] else 0


def main(argv=None):
    """
    Loads the ladder at argv[0], writes summary.json into the results root
    at argv[1], and returns the process exit code: non-zero, after naming
    the incomplete cells on stderr, when any cell is incomplete; zero
    otherwise (including when the only short cell is a not-run cold cell).
    """
    parser = argparse.ArgumentParser(description="Summarise a quarry-mcp capability ladder results tree into summary.json.")
    parser.add_argument("ladder", help="path to ladder.yaml")
    parser.add_argument("results_root", help="path to the dated results directory, e.g. results/2026-08-29")
    args = parser.parse_args(argv)

    loaded_ladder = load_ladder(args.ladder)
    summary = write_summary(loaded_ladder, args.results_root)
    if summary["incomplete"]:
        print(f"incomplete cells: {', '.join(summary['incomplete'])}", file=sys.stderr)
    return _exit_code_for_summary(summary)


if __name__ == "__main__":
    sys.exit(main())

File: bench/loomyard-eval/ladder/internal/ladder/summarize.go
// summarize.go turns a results root into the numbers a reader acts on, by re-deriving every metric
// from the raw tree rather than trusting the diagnostic-only usage.json each repetition wrote. Every
// cost metric is recomputed from that repetition's own transcript.jsonl; recall and precision -- the
// scorer's own judgment, existing nowhere else -- are read from the repetition's score.json. The
// asymmetry is the point: recomputing cost is exactly what keeping the transcripts buys, and it is
// what makes an accounting fix cost a re-report rather than a re-run.
//
// Summarize never loads the ladder file: every fact it needs about a cell -- its ladder letter, task
// id, allowed-tool list and control flag -- comes from that cell's own run.json, so a results root
// stays self-describing after the ladder file that produced it has moved or changed.

package ladder

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
)

// SummaryFile is the file name a results root's summary is written under: the single spelling of
// that name, matching how provenance.go pins ProvenanceFile and report.go pins TableFile.
const SummaryFile = "summary.json"

// costMetricNames is the fixed, ordered set of cost-metric names Summarize recomputes from every
// repetition's transcript and RenderTable renders as table columns, in the same order both use.
var costMetricNames = []string{
	"turns",
	"duration_ms",
	"cost_usd",
	"cache_read_input_tokens",
	"cache_creation_input_tokens",
	"output_tokens",
	"input_tokens_total",
	"tool_uses",
	"quarry_tool_uses",
	"grep_fallback_total",
	"tool_result_bytes",
	"read_bytes",
}

// correctnessMetricNames is the fixed set of correctness-metric names Summarize reads from each
// scored repetition's score.json, under the exact keys both scoring rules declare.
var correctnessMetricNames = []string{"recall", "precision"}

// comparisonMetricNames is every metric name a rung-vs-control comparison is eligible over: every
// cost metric plus the two correctness metrics. Unlike V1, no metric is excluded here -- this harness
// renders one identical preamble for every cell, so a grep-usage gap between a rung and its control
// carries no confound to guard against.
func comparisonMetricNames() []string {
	names := make([]string, 0, len(costMetricNames)+len(correctnessMetricNames))
	names = append(names, costMetricNames...)
	names = append(names, correctnessMetricNames...)
	return names
}

// costMetricValues reduces one repetition's recomputed Metrics to the named values costMetricNames
// declares, so summarizeCell and RenderTable read the same field-to-name mapping.
func costMetricValues(m Metrics) map[string]float64 {
	return map[string]float64{
		"turns":                       float64(m.NumTurns),
		"duration_ms":                 float64(m.DurationMS),
		"cost_usd":                    m.TotalCostUSD,
		"cache_read_input_tokens":     float64(m.CacheReadInputTokens),
		"cache_creation_input_tokens": float64(m.CacheCreationInputTokens),
		"output_tokens":               float64(m.OutputTokens),
		"input_tokens_total":          float64(m.InputTokensTotal),
		"tool_uses":                   float64(m.ToolUses),
		"quarry_tool_uses":            float64(m.QuarryToolUses),
		"grep_fallback_total":         float64(m.GrepFallbackTotal),
		"tool_result_bytes":           float64(m.ToolResultBytes),
		"read_bytes":                  float64(m.ReadBytes),
	}
}

// MetricStats is one metric's median, minimum, maximum and sample count over the repetitions that
// contributed a value to it.
type MetricStats struct {
	// Median is the metric's true median across the contributing repetitions: the middle value at an
	// odd sample count, the mean of the two middle values at an even one.
	Median float64 `json:"median"`
	// Min is the metric's minimum value across the contributing repetitions.
	Min float64 `json:"min"`
	// Max is the metric's maximum value across the contributing repetitions.
	Max float64 `json:"max"`
	// N is the count of repetitions that contributed a value to this metric. A cell's cost sample
	// count and its correctness sample count may legitimately differ, since the three exclusion rules
	// apply to correctness but not cost -- each metric's own N reflects only the repetitions that
	// actually contributed to it.
	N int `json:"n"`
}

// CellRecord is one cell's summarised results: its identity as carried by its own repetitions'
// run.json, its per-metric statistics, the three exclusion counters, and its gate-1 finding when one
// fired.
type CellRecord struct {
	// ID is the cell id.
	ID string `json:"id"`
	// Ladder is the ladder letter this cell belongs to.
	Ladder string `json:"ladder"`
	// Task is the task id this cell runs.
	Task string `json:"task"`
	// Allowed is the tool subset this cell grants.
	Allowed []string `json:"allowed"`
	// Control reports whether this cell is its ladder letter's control.
	Control bool `json:"control"`
	// Metrics maps a metric name -- every entry of costMetricNames plus correctnessMetricNames -- to
	// its summarised statistics, present only for a metric at least one repetition contributed to.
	Metrics map[string]MetricStats `json:"metrics"`
	// MaxTurnsCount is the count of this cell's repetitions that hit the turn ceiling.
	MaxTurnsCount int `json:"max_turns_count"`
	// UnscoredCount is the count of this cell's repetitions whose score record carried a false scored
	// flag for a reason other than the turn ceiling.
	UnscoredCount int `json:"unscored_count"`
	// BlindingFailedCount is the count of this cell's repetitions whose blinding-failed flag was set.
	BlindingFailedCount int `json:"blinding_failed_count"`
	// Gate1 is this cell's gate-1 (granted-tool-used) finding, nil when gate 1 did not fire.
	Gate1 *Finding `json:"gate1,omitempty"`
}

// SummaryMeta is summary.json's meta block. It names only the results root's base name -- never the
// operator-supplied path, and never a wall-clock write time -- per the overview's
// no-machine-paths-in-tracked-output decision, which is also what makes this file byte-for-byte
// golden-comparable.
type SummaryMeta struct {
	// ResultsRoot is the results root directory's own base name.
	ResultsRoot string `json:"results_root"`
}

// Comparison is one structured, mechanically checkable comparison between a rung cell and its own
// ladder letter's control, over one metric. Disjointness is non-overlapping minimum-maximum ranges;
// there is no significance testing.
type Comparison struct {
	// Cell is the rung cell's id.
	Cell string `json:"cell"`
	// Control is the control cell's id.
	Control string `json:"control"`
	// Metric is the compared metric's name.
	Metric string `json:"metric"`
	// CellMedian is the rung cell's median for this metric.
	CellMedian float64 `json:"cell_median"`
	// ControlMedian is the control cell's median for this metric.
	ControlMedian float64 `json:"control_median"`
	// CellRange is the rung cell's (min, max) pair for this metric.
	CellRange [2]float64 `json:"cell_range"`
	// ControlRange is the control cell's (min, max) pair for this metric.
	ControlRange [2]float64 `json:"control_range"`
	// Separated is true only when CellRange and ControlRange do not overlap at all.
	Separated bool `json:"separated"`
}

// RangesDisjoint reports whether the two closed ranges a and b -- each a (min, max) pair -- do not
// overlap at all. Ranges that touch at a single value -- one's max equals the other's min -- are
// treated as NOT disjoint.
func RangesDisjoint(a, b [2]float64) bool {
	return a[1] < b[0] || b[1] < a[0]
}

// Summary is the full record WriteSummary serialises to summary.json: every selected cell's
// CellRecord, every rung-vs-control comparison, the incomplete and invalid cell-id lists, and the
// meta block.
type Summary struct {
	// Cells is every selected cell's CellRecord, sorted by cell id.
	Cells []CellRecord `json:"cells"`
	// Comparisons is every rung-vs-control comparison the two cells' shared metrics support.
	Comparisons []Comparison `json:"comparisons"`
	// Incomplete lists the ids of every cell whose present-and-not-void repetition count is short of
	// the provenance record's selected cells crossed with its effective repetition count.
	Incomplete []string `json:"incomplete"`
	// Invalid lists the ids of every cell with a non-zero blinding-failed count.
	Invalid []string `json:"invalid"`
	// Meta carries the results root's base name and nothing else.
	Meta SummaryMeta `json:"meta"`
}

// Summarize walks resultsRoot's raw tree and recomputes a Summary for it. It reads each repetition's
// run.json for that repetition's cell metadata -- no ladder file is consulted -- recomputes that
// repetition's cost metrics from its own transcript.jsonl, and reads recall and precision from its
// score.json. A results root whose provenance record is absent is an error naming the missing file:
// the run subcommand writes that record before its first repetition, so a raw tree without one is a
// hand-assembled or truncated root, and reporting an empty incomplete list for it would make an
// unknowably short root read as finished.
func Summarize(resultsRoot string) (*Summary, error) {
	prov, err := ReadProvenance(resultsRoot)
	if err != nil {
		return nil, fmt.Errorf("summarize %s: %w", resultsRoot, err)
	}
	if prov == nil {
		return nil, fmt.Errorf(
			"summarize %s: missing %s -- the run subcommand writes this file before its first repetition, so a raw tree without one is a hand-assembled or truncated root",
			resultsRoot, ProvenanceFile,
		)
	}

	cellIDs := append([]string(nil), prov.SelectedCells...)
	sort.Strings(cellIDs)

	var cells []CellRecord
	var incomplete []string
	var invalid []string
	controlByLadder := map[string]CellRecord{}

	for _, cellID := range cellIDs {
		record, presentNonVoid, err := summarizeCell(resultsRoot, cellID, prov.RepsEffective)
		if err != nil {
			return nil, err
		}
		if presentNonVoid < prov.RepsEffective {
			incomplete = append(incomplete, cellID)
		}
		if record.BlindingFailedCount > 0 {
			invalid = append(invalid, cellID)
		}
		cells = append(cells, record)
		if record.Control {
			controlByLadder[record.Ladder] = record
		}
	}

	comparisons := buildComparisons(cells, controlByLadder)

	return &Summary{
		Cells:       cells,
		Comparisons: comparisons,
		Incomplete:  incomplete,
		Invalid:     invalid,
		Meta:        SummaryMeta{ResultsRoot: filepath.Base(resultsRoot)},
	}, nil
}

// summarizeCell builds cellID's CellRecord from its own repetitions, 1..repsEffective, applying the
// three exclusion rules -- all distinct: a blinding-failed repetition contributes nothing at all,
// neither cost nor correctness, is not counted present for completeness, and increments
// BlindingFailedCount; a max-turns repetition contributes cost but not recall or precision and
// increments MaxTurnsCount; a repetition whose score record carries a false scored flag for any other
// reason is likewise excluded from recall and precision and increments UnscoredCount. It also returns
// the present-and-not-void repetition count Summarize compares against repsEffective to populate the
// incomplete list.
func summarizeCell(resultsRoot, cellID string, repsEffective int) (CellRecord, int, error) {
	record := CellRecord{ID: cellID}
	values := map[string][]float64{}
	var perRepQuarryToolUses []int
	metadataCaptured := false
	presentNonVoid := 0

	for rep := 1; rep <= repsEffective; rep++ {
		dir := RepDir(resultsRoot, cellID, rep)
		state, err := ReadRunState(dir)
		if err != nil {
			// Not present: never written, or invalidated and renamed away by InvalidateRep.
			continue
		}
		if !metadataCaptured {
			record.Ladder = state.Ladder
			record.Task = state.Task
			record.Allowed = append([]string(nil), state.Allowed...)
			record.Control = state.IsControl
			metadataCaptured = true
		}
		if state.BlindingFailed {
			record.BlindingFailedCount++
			continue
		}
		presentNonVoid++

		transcript, err := readRepTranscript(dir)
		if err != nil {
			return CellRecord{}, 0, fmt.Errorf("summarize cell %s rep %d: %w", cellID, rep, err)
		}
		metrics := ComputeMetrics(transcript, state.MCPPrefix)
		for name, value := range costMetricValues(metrics) {
			values[name] = append(values[name], value)
		}
		perRepQuarryToolUses = append(perRepQuarryToolUses, metrics.QuarryToolUses)

		if state.MaxTurnsHit {
			record.MaxTurnsCount++
			continue
		}
		if !state.Scored {
			record.UnscoredCount++
			continue
		}

		score, err := readRepScore(dir)
		if err != nil {
			return CellRecord{}, 0, fmt.Errorf("summarize cell %s rep %d: %w", cellID, rep, err)
		}
		recall, recallOK := numericScoreField(score, "recall")
		precision, precisionOK := numericScoreField(score, "precision")
		if recallOK && precisionOK {
			values["recall"] = append(values["recall"], recall)
			values["precision"] = append(values["precision"], precision)
		}
	}

	record.Metrics = map[string]MetricStats{}
	for name, vs := range values {
		if len(vs) == 0 {
			continue
		}
		record.Metrics[name] = computeStats(vs)
	}

	cfg := Config{ID: cellID, Ladder: record.Ladder, Task: record.Task, Allowed: record.Allowed}
	record.Gate1 = CheckGrantedToolUsed(cfg, perRepQuarryToolUses)

	return record, presentNonVoid, nil
}

// buildComparisons builds one Comparison per (rung cell, metric) pair for every metric both the rung
// cell and its own ladder letter's control carry statistics for, in cells order and
// comparisonMetricNames order.
func buildComparisons(cells []CellRecord, controlByLadder map[string]CellRecord) []Comparison {
	var comparisons []Comparison
	for _, cell := range cells {
		if cell.Control {
			continue
		}
		control, ok := controlByLadder[cell.Ladder]
		if !ok {
			continue
		}
		for _, metric := range comparisonMetricNames() {
			cellStats, ok := cell.Metrics[metric]
			if !ok {
				continue
			}
			controlStats, ok := control.Metrics[metric]
			if !ok {
				continue
			}
			cellRange := [2]float64{cellStats.Min, cellStats.Max}
			controlRange := [2]float64{controlStats.Min, controlStats.Max}
			comparisons = append(comparisons, Comparison{
				Cell:          cell.ID,
				Control:       control.ID,
				Metric:        metric,
				CellMedian:    cellStats.Median,
				ControlMedian: controlStats.Median,
				CellRange:     cellRange,
				ControlRange:  controlRange,
				Separated:     RangesDisjoint(cellRange, controlRange),
			})
		}
	}
	return comparisons
}

// readRepTranscript parses dir/transcript.jsonl.
func readRepTranscript(dir string) (*Transcript, error) {
	path := filepath.Join(dir, TranscriptFile)
	f, err := os.Open(path)
	if err != nil {
		return nil, fmt.Errorf("read transcript %s: %w", path, err)
	}
	defer f.Close()
	t, err := ParseTranscript(f)
	if err != nil {
		return nil, fmt.Errorf("read transcript %s: %w", path, err)
	}
	return t, nil
}

// readRepScore reads and decodes dir/score.json.
func readRepScore(dir string) (ScoreRecord, error) {
	path := filepath.Join(dir, ScoreFile)
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("read score %s: %w", path, err)
	}
	var record ScoreRecord
	if err := json.Unmarshal(data, &record); err != nil {
		return nil, fmt.Errorf("parse score %s: %w", path, err)
	}
	return record, nil
}

// numericScoreField reads key out of record as a float64, reporting whether it was present and
// numeric. encoding/json decodes a JSON number into a Go float64 by default, which is the only shape
// a scorer's own reply produces for recall and precision.
func numericScoreField(record ScoreRecord, key string) (float64, bool) {
	v, ok := record[key]
	if !ok {
		return 0, false
	}
	f, ok := v.(float64)
	return f, ok
}

// computeStats reduces values, which must be non-empty, to its MetricStats: the true median, the
// minimum, the maximum and the sample count.
func computeStats(values []float64) MetricStats {
	sorted := append([]float64(nil), values...)
	sort.Float64s(sorted)
	return MetricStats{
		Median: trueMedian(sorted),
		Min:    sorted[0],
		Max:    sorted[len(sorted)-1],
		N:      len(sorted),
	}
}

// trueMedian returns the true median of sorted, which must already be sorted and non-empty: the
// middle value at an odd length, the mean of the two middle values at an even length.
func trueMedian(sorted []float64) float64 {
	n := len(sorted)
	mid := n / 2
	if n%2 == 1 {
		return sorted[mid]
	}
	return (sorted[mid-1] + sorted[mid]) / 2
}

// WriteSummary writes s to resultsRoot/summary.json.
func WriteSummary(resultsRoot string, s *Summary) error {
	data, err := json.MarshalIndent(s, "", "  ")
	if err != nil {
		return fmt.Errorf("write summary: %w", err)
	}
	path := filepath.Join(resultsRoot, SummaryFile)
	if err := os.WriteFile(path, data, 0o644); err != nil {
		return fmt.Errorf("write summary %s: %w", path, err)
	}
	return nil
}
